README.md

# pyDE1

## License

Copyright © 2021 Jeff Kletsky. All Rights Reserved.

License for this software, part of the pyDE1 package, is granted under

GNU General Public License v3.0 only

SPDX-License-Identifier: GPL-3.0-only


## Overview

A fully functional, API-first implementation of core software for control
and use of the Decent Espresso DE1.

It provides core components, which can be run as an unattended service
that automatically starts at boot, to supply stable, versioned APIs
to provide all primary functions of use of a DE1 and data collection around it.

A web app has been able to demonstrate sufficiency of the APIs
and functionality for the majority of day-to-day operations,
including real-time graphing and history display. A running example
is available, linked from DecentForum.com. It uses the supplied,
stand-alone "replay" application to play back a shot in real time.
This tool is also useful for development of companion applications.

Profiles and real-time data are captured into a SQLite3 database
that allows multiple, concurrent access.

A stand-alone program is provided that can automatically upload "shots"
to Visualizer as soon as they complete. It can also notify consumers
of the URL returned.

An example program is provided that generates legacy-style, "shot files" that
are compatible with Visualizer and John Weiss' shot-plotting programs.

The APIs are under semantic versioning. The REST-like, HTTP-transport
versions can be retrieved from `version` at the document root, and
also include the Python and package versions installed. Each of the
JSON-formatted, MQTT packets contains a `version` key:value for that
payload.

Consumers of these APIs should only need to understand high-level
actions, such as "Here is a profile blob, please load it." The
operations and choice of connectivity to the devices is "hidden"
behind the APIs.


## Revision History

See also CHANGELOG.md

* 2021-10-31 – 0.9.0 Update for beta (functionally complete)
* 2021-09-28 – 0.8.0 Implementation as unattended services
* 2021-08-12 – 0.7.0 sets profiles by ID, auto-reconnect, replay, uploader
* 2021-07-25 – 0.6.0 adds database store
* 2021-07-14 – 0.5.0, "worked example" description
* 2021-07-03 – Updated for release 0.4.0, see also CHANGELOG.md
* 2021-06-26 – Content and organizational updates for release 0.3.0
* 2021-06-22 – Updated for release 0.2.0
* 2021-06-11 – Updated for release 0.1.0
* 2021-06-08 – Initial release

## Support and Discussion

Support and discussion is active at DecentForum.com, on Discord in the
Decent Espresso server and, to some extent, on the Espresso
Aficianados server in the Manufacturers: decent channel. Support is,
unfortunately, ***not*** available through Decent Diaspora on
Basecamp.

Thanks to all that have been trying this out and providing valuable feedback!

See also
[https://github.com/jeffsf/pyDE1](https://github.com/jeffsf/pyDE1)
where the *beta* branch is now current.

The alpha branch will be retained there as some of the remaining development
tooling and examples will no longer be actively maintained. They will be
removed from the beta and future release branches.

## What's New

_**Please see CHANGELOG.md for more details**_

## 0.9.0 – 2021-10-31

### Overview

Version 0.9.0 represents the first beta release of this project. "Beta" here
represents a product that is believed to be functionally complete
for the intended scope and has undergone testing.


### New

* The flush-control features of *experimental* Firmware 1283 were implemented
  and include control of target duration, temperature, and flow. 46c0481

* Clean, Descale, and Transport functionality is now available through the API.
  65f2ac9

* Provide asynchronous firmware upload through API. d6a2dbc, 32436a9

* GET of DE1_STATE enabled. 2b4435e

* Rewrite of logging and logging configuration. "Early" logging is captured
  and routed to the log file, once it is opened. Log levels and formatters
  can be easily configured through the YAML config files. b759168, 39c714d,
  7df0397, d3e128c

* Provide logging over MQTT for client use (in addition to console and
  log file). 019bed0

* Profile frame logging provides "not" names for unset FrameFlags to clarify
  log messages. For example, the absence of `CtrlF` is now rendered as `CtrlP`.
  c842565

* MQTT "Will" implemented, reporting unexpected MQTT disconnects. 22d06b4

* Feature flags have been added to formalize access to DE1 and firmware
  abilities. d7405b0


### Fixed

* Loop-level, exception-initiated shutdowns now terminate more cleanly. 0b593d0

* An error condition when no scale was present during a "shot"
  has been resolved. ffae2f

* An error condition when a DE1 connected and the profile was not yet known
  has been resolved 58bbfad

* AutoTareNotification and StopAtNotification now populate sender. 9f39d08

* A very early termination of the program (before processes are defined) now
  terminates more cleanly. 4f95c34

* Reset the scale history if a gap in reports is too long, such as
  from a disconnect-reconnect sequence. 48a35ca

*See CHANGELOG.md for other changes and removals*

## Requirements

Python 3.8 or later.

Current plans are to continue to support two versions prior to
the latest Python version. [PEP 664](https://www.python.org/dev/peps/pep-0664/)
shows a schedule of October, 2022 for Python 3.11.0 release.
Users should plan on moving to at least Python 3.9 prior to that release.
Python 3.9 should be "standard" with the forthcoming, Raspberry Pi OS release
based on Debian Bullseye. Upstream Debian Bullseye is already available.

Available through `pip`:

* `bleak`
* `aiosqlite`
* `paho-mqtt`
* `PyYAML`
* `requests`

An MQTT broker compatible with MQTT 5 clients, such as `mosquitto 2.0`

A production-quality web server, such as `nginx` is highly recommended
operating as a reverse proxy, rather than directly exposing the Python
web server.

## Short-Term Priorities

* Stand-alone documentation
* Quick-start guide (awaiting release of Raspberry OS on Debian Bullseye)

## Out of Scope

* Single-command read of the DE1 debug register
* The private, "calibration" CUUID commands have not been "unlocked".


## Related Work (Other Projects)

* Componentize JavaScript real-time graph rendering
* Develop GraphQL access to database


## Status — Beta

This code is used on a daily basis for operation of the author's DE1.

This code is supplied under GPL v3.0 and is subject to the terms
of that license, including limitation of liability.


CHANGELOG.md

# Changelog

## 0.9.0 – 2021-10-31

### Overview

Functionality for the beta release completed and tested.

### New

* The flush-control features of *experimental* Firmware 1283 were implemented
  and include control of target duration, temperature, and flow. 46c0481

* Clean, Descale, and Transport functionality is now available through the API.
  65f2ac9

* Provide asynchronous firmware upload through API. d6a2dbc, 32436a9

* GET of DE1_STATE enabled. 2b4435e

* Rewrite of logging and logging configuration. "Early" logging is captured
  and routed to the log file, once it is opened. Log levels and formatters
  can be easily configured through the YAML config files. b759168, 39c714d,
  7df0397, d3e128c, cabab97

* Provide logging over MQTT for client use (in addition to console and
  log file). 019bed0

* Profile frame logging provides "not" names for unset FrameFlags to clarify
  log messages. For example, the absence of `CtrlF` is now rendered as `CtrlP`.
  c842565

* MQTT "Will" implemented, reporting unexpected MQTT disconnects. 22d06b4

* Feature flags have been added to formalize access to DE1 and firmware
  abilities. d7405b0


### Fixed

* Loop-level, exception-initiated shutdowns now terminate more cleanly. 0b593d0

* An error condition when no scale was present during a "shot"
  has been resolved. ffae2f

* An error condition when a DE1 connected and the profile was not yet known
  has been resolved 58bbfad

* AutoTareNotification and StopAtNotification now populate sender. 9f39d08

* A very early termination of the program (before processes are defined) now
  terminates more cleanly. 4f95c34

* ScaleProcessor: Reset the history if a gap in reports is too long, such as
  from a disconnect-reconnect sequence. 48a35ca



### Changed

* `c_api` was updated with new information. 46c0481

* The firmware version is read early in the DE1 initialization to determine
  the range of valid MMRs and how to efficiently read them. 46c0481

* The `ModeControl` class was refactored into `flow_sequencer`. 46c0481

* MMRs that are not able to be decoded (such as not implemented), are logged
  along with the value received. 2d0fa24

* Return 400 Bad Request for PATCH/PUT with no content. d00bd24

* Change MQTT to not request retaining messages from pyDE1. 8a8ba5e

* Logging level and wording changes. 99ec22f, b31c850

* **MAPPING 4.0.0** rewrites `IsAt` to use an enum, rather than the class
  to define the target, simplifying package inclusion. 78cea85

* Rework imports to remove order dependencies and simplify. c895f7d, b31c850

* Improve reconnection algorithm for DE1 and Scale. 6be3e5a

* Improve camelcase_from_underscore(). 0b40fe9

* Do not try to reconnect DE1 or Scale while shutting down. bd21a93

* Inbound (HTTP) API: Check DE1 and scale is_ready instead of is_connected.
  5de28e7


### Deprecated

### Removed

* Remove unused Config.set_logging(). 2b104e6

* Remove feature.py as previously incorporated into FeatureFlag. 469ee96




## 0.8.0 – 2021-09-28

### Overview

This release focused on converting command-line executables to robust,
self-starting, and supervised services. Both the core pyDE1 controller
and the Visualizer uploader now can be started with `systemd`
automatically at boot. Configuration of many parameters can be done
through YAML files (simple, human-friendly syntax), by default in
`/usr/local/pyde1/`. Command-line parameters, usable by the service unit files,
can be used to override the config-file location.

Logging configuration may change prior to "beta". At this time it is only
configurable in the output format and level for the *stderr* and *file* loggers.  
By default, the *stderr* logger is at the WARNING level abd without timestamps,
as it is managed through `systemd` when being run as a service. A command-line
parameter allows for timestamped output at the DEBUG level for interactive use.

### New

* Services run under `systemd`
    * Service ("unit") files for `pyde1.service` and `pyde1-visualizer.service`
    * Config files in YAML form
* Auto-off, configurable
* Track the IDs of connected Bluetooth devices for cleanup under Linux and
    disconnect them at the Bluez level in the case of a non-graceful exit
* MQTT supports authorization and access-control lists
* Visualizer: Don't upload short "shots", such as for flushing (configurable)
* Stop-at-weight offset configurable through `pyde1.conf`
* Database:
    * Self-initialize, if needed
    * Check for the proper schema at start
* Replay: config file and command-line switches allow easier configuration,
    including sequence ID and MQTT topic root


### Fixed

* MQTT (outbound) API will now detect connection or authentication failures
    with the broker and terminate pyDE1
* FlowSequencer no longer raises exception when trying to report that
    the steam time is not managed directly by the software.
    (It is managed by the DE1 firmware.)
* Mass-flow estimates had an off-by-one error that was corrected
* Replay now properly reports sequence_id on gate notifications


### Changed

* Paths changed to `/var/log/pyde1` and `/var/lib/pyde1/pyde1.sqlite`
    by default (configurable)
* Refactored and unified shutdown processes
    * **NB: SIGHUP is no longer used for log rotation. It is a termination signal.**
* Refactored supervised processes to handle uncaught exceptions and
    properly terminate for automated restart
* Visualizer: log to `pyde1-visualizer.log` by default
* Stop-at-weight internally includes 170 ms to account for the "fall-time"
    from the basket to the cup.
* Logging:
    * Switched to a file-watcher handler so that log rotation should
        be transparent, without the need of a signal
    * Provide better control of formatting and level for use with `systemd`
        (service) infrastructure
    * Change default file name to `pyde1.log`
    * Add `--console` command-line flag to provide timestamped,
        DEBUG-level output to assist in development and debugging

src/pyDE1/config.py
"""
Copyright © 2021 Jeff Kletsky. All Rights Reserved.

License for this software, part of the pyDE1 package, is granted under
GNU General Public License v3.0 only
SPDX-License-Identifier: GPL-3.0-only

Deciding no how to handle "global" config is a bit more complicated
for multiprocessing than it is for a single process (with or without threads).

Since each process loads Python, to get the same config values for all processes
they all need to either read the same config file, or create copies of the
"original" for themselves. Using something like a multiprocessing.Manager()
seems excessive, as it is unlikely that the values will change during a run.
Leaving that open for the future makes some sense.

That means that each process will need to be either passed the name of the
config file to be read, or the results of that (and any changes prior to
creation of the child processes). Passing the config instance seems to be
no more complex than passing the file reference.

The config data should be easy to access in each of the modules.
Several options for "singleton" behavior:

1) Use a bare module

Advantages:

    Simple access

        import pyDE1.config as config
        do_something_with(config.section.VALUE)

    Only one instance "guaranteed"

Disadvantages:

    No top-level properties

        Though there may be a way to define them, the @property decorator
        will create a property object, but it doesn't seem useful.

        This is one of the main reasons that the Singleton object is used
        for the primary components of pyDE1 (the other is no dangling instances)


2) Module that defines an instance, refer to the instance

Advantages:

    Simple access

        from pyDE1.config import config
        do_something_with(config.section.VALUE)

Disadvantages:

    It is possible that second instance, disconnected from the "real" one,
    could be generated with config = Config()

3) Subclass Singleton

Advantages:

    Only one instance "guaranteed"

Disadvantages:

    More complex access

        from pyDE1.config import Config
        pointer_to_config = Config()    # At the module or function level
        do_something_with(config.section.VALUE)



"""

import logging
from typing import Optional

import pyDE1
from pyDE1.config_load import ConfigYAML, ConfigLoadable
from pyDE1.pyde1_logging import (
    ConfigLogging, ConfigLoggingFormatters, ConfigLoggingHandlers
)

DEFAULT_CONFIG_FILE = '/usr/local/etc/pyde1/pyde1.conf'

logger = pyDE1.getLogger('Config')


class Config (ConfigYAML):

    DEFAULT_CONFIG_FILE = '/usr/local/etc/pyde1/pyde1.conf'

    def __init__(self):
        super(Config, self).__init__()
        self.bluetooth = _Bluetooth()
        self.database = _Database()
        self.de1 = _DE1()
        self.http = _HTTP(self)    # Calculating timeout needs bluetooth
        self.logging = _Logging()
        self.mqtt = _MQTT()


# This craziness is so pyCharm autocompletes
# Otherwise typing.SimpleNamespace() would be sufficient


class _MQTT (ConfigLoadable):
    def __init__(self):
        self.TOPIC_ROOT = 'pyDE1'
        self.CLIENT_ID_PREFIX = 'pyde1'
        self.BROKER_HOSTNAME = '::1'
        self.BROKER_PORT = 1883
        self.TRANSPORT = 'tcp'
        self.TLS_CONTEXT = None
        self.KEEPALIVE = 60
        self.USERNAME = None
        self.PASSWORD = None
        self.DEBUG = False


class _HTTP (ConfigLoadable):
    def __init__(self, parent):
        self.SERVER_HOST = ''
        self.SERVER_PORT = 1234
        self.SERVER_ROOT = '/'
        # adaptive_allonge.json is 7632 bytes
        self.PATCH_SIZE_LIMIT = 16384
        # Seconds, before abandoning the request
        self.ASYNC_TIMEOUT = 1.0
        # Seconds, 20*2 frames + head + tail at ~100 ms each
        self.PROFILE_TIMEOUT = 4.5
        self.FIRMWARE_TIMEOUT = 15  # Seconds for upload and start (~260 kbps)
        self._response_timeout = None

        # If true, don't output nodes that have no value (write-only)
        # or are empty dicts
        # Otherwise math.nan fills in for the missing value
        # As not compliant with RFC 7159, some parsers may fail with NaN
        # although it is permitted by ECMAScript and JavaScript
        # A False setting is intended to be a development/exploration tool
        # This feature be considered as deprecated
        self.PRUNE_EMPTY_NODES = True

        self._parent = parent   # Path to get to bluetooth

    @property
    def RESPONSE_TIMEOUT(self):
        # See pyDE1/dispatcher/implementation.py
        # Right now, single timeout, bounded by scan/connect
        # This is in addition to the timeout in the implementation
        if self._response_timeout:
            retval = self._response_timeout
        else:
            retval = (max((self._parent.bluetooth.SCAN_TIME
                           + self._parent.bluetooth.CONNECT_TIMEOUT
                           + self.ASYNC_TIMEOUT),
                          (self.PROFILE_TIMEOUT
                           + self.ASYNC_TIMEOUT
                           + self._parent.de1.CUUID_LOCK_WAIT_TIMEOUT))
                      + 0.100)
        return retval

    @RESPONSE_TIMEOUT.setter
    def RESPONSE_TIMEOUT(self, value):
        self._response_timeout = value


class _Logging (ConfigLogging):
    def __init__(self):
        super(_Logging, self).__init__()


class _LoggingFormatters (ConfigLoggingFormatters):
    def __init__(self):
        super(_LoggingFormatters, self).__init__()


class _LoggingHandlers (ConfigLoggingHandlers):
    def __init__(self):
        super(_LoggingHandlers, self).__init__()


class _Bluetooth (ConfigLoadable):
    def __init__(self):
        self.SCAN_TIME = 5  # Seconds
        self.CONNECT_TIMEOUT = 10  # Seconds
        self.DISCONNECT_TIMEOUT = 5  # Seconds
        self.SCAN_CACHE_EXPIRY = 300  # Seconds, probably too long
        self.RECONNECT_RETRY_COUNT = 10 # Before using RECONNECT_GAP
        self.RECONNECT_GAP = 10 # Seconds
        # Files that hold the Bluetooth ID of connected devices
        # for potential cleanup by supervisor scripts
        self.ID_FILE_DIRECTORY = '/var/lib/pyde1/'
        self.ID_FILE_SUFFIX = '.btid'


class _Database (ConfigLoadable):
    def __init__(self):
        self.FILENAME = '/var/lib/pyde1/pyde1.sqlite3'
        self.BACKUP_TIMEOUT = 60  # seconds
        self.BACKUP_COMPRESSION_EXECUTABLE = 'xz'


class _DE1 (ConfigLoadable):
    def __init__(self):
        self.LINE_FREQUENCY = 60
        self.MAX_WAIT_FOR_READY_EVENTS = 3.5 # Seconds (25 at 0.1 each)
        self.CUUID_LOCK_WAIT_TIMEOUT = 2 # Seconds
        # Do these "settings" belong here,
        # or should they be separated from parameters?
        self.DEFAULT_AUTO_OFF_TIME = None   # Minutes
        self.STOP_AT_WEIGHT_ADJUST = -0.07  # Secs, larger increases weight


config = Config()